rpki-validator/src/main/java/net/ripe/rpki/validator3/api/bgp/BgpPreviewService.java

```diff
@@ -238,16 +229,82 @@ public static BgpPreviewEntry of(Asn origin, IpRange prefix, Validity validity)
         }
 
+        public abstract IpRange getPrefix();
+        abstract BgpPreviewEntry ofValidity(Validity validity);
+
+        public Asn getOrigin() {
+            return new Asn(Integer.toUnsignedLong(origin));
+        }
+
+    }
+
+    @lombok.Value
+    @lombok.EqualsAndHashCode(callSuper = true)
+    @lombok.ToString(callSuper = true)
+    public static class BgpPreviewEntry4 extends BgpPreviewEntry {
+        short prefixLength;
+        int prefix;
+
+        private BgpPreviewEntry4(int origin, Validity validity, short prefixLength, int prefix) {
+            super(origin, validity);
+            this.prefixLength = prefixLength;
+            this.prefix = prefix;
+        }
+
+        public static BgpPreviewEntry4 of(Asn origin, Validity validity, IpRange prefix) {
+            return new BgpPreviewEntry4((int) origin.longValue(), validity, (short) prefix.getPrefixLength(), (int) ((Ipv4Address) prefix.getStart()).longValue());
+        }
+
+        public IpRange getPrefix() {
+            return IpRange.prefix(new Ipv4Address(Integer.toUnsignedLong(prefix)), prefixLength);
+        }
+
         BgpPreviewEntry ofValidity(Validity validity) {
-            return new BgpPreviewEntry(origin, prefix, validity);
+            return new BgpPreviewEntry4(origin, validity, prefixLength, prefix);
         }
 
-        public Asn getOrigin() {
-            return new Asn(origin);
+    }
+
+    @lombok.Value
+    @lombok.EqualsAndHashCode(callSuper = true)
+    @lombok.ToString(callSuper = true)
+    public static class BgpPreviewEntry6 extends BgpPreviewEntry {
+        private static BigInteger TWO_64 = BigInteger.ONE.shiftLeft(64);
+        private static BigInteger TWO_64_MINUS_1 = TWO_64.subtract(BigInteger.ONE);
+
+        short prefixLength;
+        long prefixHi;
+        long prefixLo;
+
+        private BgpPreviewEntry6(int origin, Validity validity, short prefixLength, long prefixHi, long prefixLo) {
+            super(origin, validity);
+            this.prefixLength = prefixLength;
+            this.prefixHi = prefixHi;
+            this.prefixLo = prefixLo;
+        }
+
+        public static BgpPreviewEntry6 of(Asn origin, Validity validity, IpRange prefix) {
+            BigInteger value = prefix.getStart().getValue();
+            short prefixLength = (short) prefix.getPrefixLength();
+            long prefixHi = value.shiftRight(64).and(TWO_64_MINUS_1).longValue();
+            long prefixLo = value.and(TWO_64_MINUS_1).longValue();
+            return new BgpPreviewEntry6((int) origin.longValue(), validity, prefixLength, prefixHi, prefixLo);
         }
 
         public IpRange getPrefix() {
-            return prefix.toIpRange();
+            BigInteger prefix = BigInteger.valueOf(prefixHi);
+            if (prefixHi < 0) {
+                prefix = prefix.add(TWO_64);
+            }
+            prefix = prefix.shiftLeft(64).add(BigInteger.valueOf(prefixLo));
+            if (prefixLo < 0) {
```

Comment:
My two complement's rusty: Would be easier for me to read when you perform the `add(TWO_64) first (if needed) and then add if to the shifted prefix.

As a bonus this may be negibly faster since `prefixLo` is shorter than prefix post-shifting.